### PR TITLE
feat: [resync] add `backing_action` table

### PR DESCRIPTION
**Pull request by sk-yagi**

### CODE REVIEW

File: src/indexers/chain/backing.ts [hunk 1]
```diff
@@ -121,43 +157,76 @@ export const event = $.event(
...
 
 export const rollback = $.rollback(
   async ({ connections: { sql, views }, point }) => {
+    // Carefully check this!
+    await sql`
+      DELETE FROM chain.backing b
+      USING chain.output o
+      WHERE o.id = b.id AND o.spent_slot > ${slotFrom(point)}
+    `;
+
```
Comment (sk-yagi) [inline src/indexers/chain/backing.ts:226]: Delete this

File: src/indexers/chain/backing.ts [hunk 2]
```diff
@@ -25,6 +26,23 @@ export type ChainBacking = {
   unbackedAt: UnixTime | null;
 };
 
+const AllAction = {
+  0: "back",
+  1: "unback",
+  2: "claim_reward", // to be supported
+  3: "migrate", // to be supported
+};
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:34]: Create a custom type https://github.com/teiki-network/teiki-index/blob/main/src/indexers/chain/project.ts#L63

Comment (sk-shishi) [reply src/indexers/chain/backing.ts:34]:
```ts
const AllActions = ["back", "unback", "claim_rewards", "migrate"] as const;

export type ChainBackingAction = {
  // ...
  action: (typeof AllActions)[number]
}
```

Then use `AllActions` for creating a SQL type.

File: src/indexers/chain/backing.ts [hunk 3]
```diff
@@ -40,9 +58,7 @@ export const setup = $.setup(async ({ sql }) => {
       backing_amount bigint NOT NULL,
       milestone_backed smallint NOT NULL,
       backing_message text,
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:60]: Drop this

File: src/indexers/chain/backing.ts [hunk 4]
```diff
@@ -53,6 +69,18 @@ export const setup = $.setup(async ({ sql }) => {
...
+      id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+      project_id varchar(64) NOT NULL,
+      actor_address text NOT NULL,
+      amount bigint NOT NULL,
+      message text,
+      time timestamptz NOT NULL,
+      slot integer NOT NULL,
+      action smallint
+    )
+  `;
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:83]: Create a few indexes for this table. At least: `project_id`, `actor_address`, `action`, `slot` and maybe `time` also. 
And where is my `tx_id`?

File: src/indexers/chain/backing.ts [hunk 5]
```diff
@@ -84,34 +112,59 @@ export const filter = $.filter(
...
+      INNER JOIN chain.output o ON b.id = o.id
+      WHERE (o.tx_id, o.tx_ix) IN ${sql(
+        tx.body.inputs.map((input) => sql([input.txId, input.index]))
+      )}
+      GROUP BY
+        (b.project_id, b.backer_address)
     `;
+
+    // BackingTupleKey
+    const actionRef: Map<string, BackingDiff> = new Map<string, BackingDiff>();
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:156]:
```suggestion
    const actionRef = new Map<string, {back: Lovelace; unback: Lovelace}>();
```

File: src/indexers/chain/backing.ts [hunk 6]
```diff
@@ -84,34 +112,59 @@ export const filter = $.filter(
...
+    unbackResult.forEach((value) => {
+      const key: BackingTupleKey = [value.projectId, value.actorAddress];
+      actionRef.set(toJson(key), {
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:159]: It's a throw-away key.

```suggestion
      const key = `${value.projectId}|${value.actorAddress}`;
      actionRef.set(key, {
```

File: src/indexers/chain/backing.ts [hunk 7]
```diff
@@ -84,34 +112,59 @@ export const filter = $.filter(
   }
 );
 
+type BackingDiff = {
+  unback: Lovelace;
+  back: Lovelace;
+};
+
+type UnbackResult = {
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:120]: This is a throw-away type.

File: src/indexers/chain/backing.ts [hunk 8]
```diff
@@ -84,34 +112,59 @@ export const filter = $.filter(
...
+    const unbackResult = await sql<UnbackResult[]>`
+      SELECT
+        SUM(b.backing_amount) AS total_amount,
+        b.project_id,
+        b.backer_address AS actor_address
+      FROM chain.backing b
+      INNER JOIN chain.output o ON b.id = o.id
+      WHERE (o.tx_id, o.tx_ix) IN ${sql(
+        tx.body.inputs.map((input) => sql([input.txId, input.index]))
+      )}
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:150]: Help me leave a TODO here. Just tag me, don't say anything.

File: src/indexers/chain/backing.ts [hunk 9]
```diff
@@ -121,43 +174,69 @@ export const event = $.event(
...
+
+        const backerAddress = deconstructAddress(
+          lucid,
+          backingDatum.backerAddress
+        );
+        const projectId = backingDatum.projectId.id;
+        const tupleKey: BackingTupleKey = [projectId, backerAddress];
+        const res = actionRef.get(toJson(tupleKey));
+
+        if (res) {
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:186]:
```ts
let ref = actionRef.get(key);

if (ref == null) {
  ref = {unback: 0, back: 0};
  actionRef.set(key, ref);
}

// Mutate ref here
ref.back += output.value.lovelace;
```

And add a
```ts
const backingAmount = output.value.lovelace;
```

File: src/indexers/chain/backing.ts [hunk 10]
```diff
@@ -121,43 +174,69 @@ export const event = $.event(
...
+            backingMessage: message,
+            backedAt: txTimeStart,
           },
         ];
       });
       if (!backings.length) console.warn("there is no valid backing");
       else await sql`INSERT INTO chain.backing ${sql(backings)}`;
     }
+
+    for (const [_key, value] of actionRef) {
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:216]: 1. Use `.entries()`.
2. Don't add the `_` prefix, you can call it `rawKey` though.

File: src/indexers/chain/backing.ts [hunk 11]
```diff
@@ -84,34 +112,59 @@ export const filter = $.filter(
...
-      WHERE
-        o.id = b.id
-        AND (o.tx_id, o.tx_ix) IN ${sql(
-          tx.body.inputs.map((input) => sql([input.txId, input.index]))
-        )}
+    const message = extractCip20Message(tx)?.join("\n") || null;
+
+    const unbackResult = await sql<UnbackResult[]>`
+      SELECT
+        SUM(b.backing_amount) AS total_amount,
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:143]:
```suggestion
        SUM(b.backing_amount)::bigint AS total_amount,
```

File: src/indexers/chain/backing.ts [hunk 12]
```diff
@@ -53,6 +63,51 @@ export const setup = $.setup(async ({ sql }) => {
...
+    CREATE TABLE IF NOT EXISTS chain.backing_action (
+      id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+      action chain.backing_all_actions NOT NULL,
+      project_id varchar(64) NOT NULL,
+      actor_address text NOT NULL,
+      amount bigint NOT NULL,
+      time timestamptz NOT NULL,
+      message text,
+      slot integer NOT NULL,
+      tx_id text NOT NULL
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:88]: Just for consistency `varchar(64)`

File: src/indexers/chain/backing.ts [hunk 13]
```diff
@@ -53,6 +63,51 @@ export const setup = $.setup(async ({ sql }) => {
...
+  await sql`
+    CREATE TABLE IF NOT EXISTS chain.backing_action (
+      id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
+      action chain.backing_all_actions NOT NULL,
+      project_id varchar(64) NOT NULL,
+      actor_address text NOT NULL,
+      amount bigint NOT NULL,
+      time timestamptz NOT NULL,
+      message text,
+      slot integer NOT NULL,
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:87]: If we use `NOT NULL REFERENCES chain.block (slot) ON DELETE CASCADE` then we don't need the custom rollback anymore.

File: src/indexers/chain/backing.ts [hunk 14]
```diff
@@ -84,34 +139,50 @@ export const filter = $.filter(
   }
 );
 
+// type BackingTupleKey = [Hex, Address];
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:142]: Remove?

File: src/indexers/chain/backing.ts [hunk 15]
```diff
@@ -84,34 +139,50 @@ export const filter = $.filter(
...
+      INNER JOIN chain.output o ON b.id = o.id
+      WHERE (o.tx_id, o.tx_ix) IN ${sql(
+        tx.body.inputs.map((input) => sql([input.txId, input.index])) // @sk-shishi
+      )}
+      GROUP BY
+        (b.project_id, b.backer_address)
     `;
+
+    const actionRef: Map<string, { unback: Lovelace; back: Lovelace }> =
+      new Map<string, { unback: Lovelace; back: Lovelace }>();
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:174]:
```suggestion
    const actionRef = new Map<string, { unback: Lovelace; back: Lovelace }>();
```

File: src/indexers/chain/backing.ts [hunk 16]
```diff
@@ -121,43 +192,61 @@ export const event = $.event(
...
+        slot,
+        txId: tx.id,
+      };
+      await sql`INSERT INTO chain.backing_action ${sql(backerAction)}`;
+    }
     driver.refresh("views.project_summary");
   }
 );
 
 export const rollback = $.rollback(
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:245]: We can drop this.

File: src/indexers/chain/backing.ts [hunk 17]
```diff
@@ -121,45 +193,53 @@ export const event = $.event(
...
-      WHERE
-        o.id = b.id
-        AND o.spent_slot > ${slotFrom(point)}
-    `;
-    views.refresh("views.project_summary");
+    for (const [rawKey, value] of actionRef.entries()) {
+      const [projectId, actorAddress] = rawKey.split("|");
+      const delta = value.back - value.unback;
+      const action = delta > 0n ? "back" : "unback";
+      const backerAction: ChainBackingAction = {
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:230]:
```suggestion
      const backingAction: ChainBackingAction = {
```
xD

File: src/indexers/chain/backing.ts [hunk 18]
```diff
@@ -53,6 +63,55 @@ export const setup = $.setup(async ({ sql }) => {
...
+  await sql`
+    CREATE INDEX IF NOT EXISTS backing_action_slot_index
+      ON chain.backing_action(slot)
+  `;
+  await sql`
+    CREATE INDEX IF NOT EXISTS backing_action_time_index
+      ON chain.backing_action(time)
+  `;
+  await sql`
+    CREATE INDEX IF NOT EXISTS backing_action_txid_index
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:112]:
```suggestion
    CREATE INDEX IF NOT EXISTS backing_action_tx_id_index
```

File: src/indexers/chain/backing.ts [hunk 19]
```diff
@@ -53,6 +63,55 @@ export const setup = $.setup(async ({ sql }) => {
     CREATE INDEX IF NOT EXISTS backing_backer_address_index
       ON chain.backing(backer_address)
   `;
+
+  await sql`
+    DO $$ BEGIN
+      IF to_regtype('chain.backing_all_actions') IS NULL THEN
+        CREATE TYPE chain.backing_all_actions AS ENUM (${sql.unsafe(
```
Comment (sk-shishi) [inline src/indexers/chain/backing.ts:70]:
```suggestion
      IF to_regtype('chain.backing_action_type') IS NULL THEN
        CREATE TYPE chain.backing_action_type AS ENUM (${sql.unsafe(
```

**Review by sk-shishi — APPROVED** [review]

**Review by sk-shishi — APPROVED** [review]
LGTM

**Pull request merged**